scripts/nbconverted/3_add_batch_effects.py

```diff
@@ -0,0 +1,214 @@
+
+# coding: utf-8
+
+# # Add batch effects
+# 
+# Say we are interested in identifying genes that differentiate between disease vs normal states.  However our dataset includes samples from different tissues or time points and there are variations in gene expression that are due to these other conditions and do not have to do with disease state.  These non-relevant variations in the data are called *batch effects*.  
+# 
+# We want to model these batch effects.  To do this we will:
+# 1. Partition our simulated data into n batches
+# 2. For each partition we will randomly shift the expression data.  We randomly generate a binary vector of length=number of genes (*offset vector*).  This vector will serve as the direction that we will shift to.  Then we also have a random scalar that will tell us how big of a step to take in our random direction (*stretch factor*).  We shift our partitioned data by: batch effect partition = partitioned data + stretch factor * offset vector
+# 3. Repeat this for each partition
+# 4. Append all batch effect partitions together
+# 
+
+# In[1]:
+
+
+get_ipython().run_line_magic('load_ext', 'autoreload')
+get_ipython().run_line_magic('autoreload', '2')
+
+import os
+import pandas as pd
+import numpy as np
+import random
+import glob
+import umap
+
+from ggplot import *
+from sklearn.decomposition import PCA
+from numpy.random import seed
+randomState = 123
+seed(randomState)
+
+
+# In[2]:
+
+
+# Parameters
+analysis_name = 'treatment'
+num_simulations = 10
+num_batches = [1,2,3,4,5,6,7,8,9,10,15,20,25,30,35,40,45,50]
```

Comment:
It may be useful to also think about nested batches for future directions. In realistic settings, 50 randomly spaced batches are less likely than a few big batches with subtle data clusters within each, for example.

Reply:
Along the same line of thought, playing with different level of correlations between batches may be interesting.

Reply:
Good points.  Casey and I have discussed this as a future direction once we are able to get realistic simulated data generated.